lib/std/meta.zig

```diff
@@ -1094,6 +1094,47 @@ test "sizeof" {
     testing.expect(sizeof(c_void) == 1);
 }
 
+pub const CIntLiteralRadix = enum { decimal, octal, hexadecimal };
+
+fn PromoteIntLiteralReturnType(comptime SuffixType: type, comptime target: comptime_int, comptime radix: CIntLiteralRadix) type {
+    const signed_decimal = [_]type{ c_int, c_long, c_longlong };
+    const signed_oct_hex = [_]type{ c_int, c_uint, c_long, c_ulong, c_longlong, c_ulonglong };
+    const unsigned = [_]type{ c_uint, c_ulong, c_ulonglong };
+
+    const list: []const type = if (@typeInfo(SuffixType).Int.signedness == .unsigned)
+        &unsigned
+    else if (radix == .decimal)
+        &signed_decimal
+    else
+        &signed_oct_hex;
+
+    var pos = mem.indexOfScalar(type, list, SuffixType).?;
+
+    while (pos < list.len) : (pos += 1) {
+        if (target >= math.minInt(list[pos]) and target <= math.maxInt(list[pos])) {
+            return list[pos];
+        }
+    }
+    @compileError("Integer literal does not fit in compatible types");
```

Comment:
According to the C spec we can put these in an "extended integer type" so `{iu}128` or maybe even `comptime_int`?

Reply:
`__int128_t` is not an "extended integer type" because `intmax_t` is still 64 bit.
https://stackoverflow.com/questions/57332208/example-of-an-extended-integer-type

However, a quick test revealed that gcc issues a warning, but still upgrades to __int128_t:
```c
#include <stdint.h>
#include <stdio.h>

int main(){
    printf("%ld\n", sizeof(typeof(21342323423423234234234)));
    printf("sizeof(__intmax_t): %ld\n", sizeof(__intmax_t));
}
```
```
$ gcc a.c
a.c: In function ‘main’:
a.c:5:35: warning: integer constant is too large for its type
    5 |     printf("%ld\n", sizeof(typeof(21342323423423234234234)));
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~
$ ./a.out 
16
sizeof(__intmax_t): 8

```
clang and MSVC produce an error.